### ISSUE TITLE
Fix video scrubbing

### ISSUE THREAD

**Issue by KianBadie** [caltechlibrary/cell-atlas#39]
Video scrubbing flashes incorrect frames when scrubbing sometimes. This issue is for fixing that bug and/or implementing new ways of scrubbing.

**Comment by KianBadie** [comment]
@coiko I just wanted to confirm again with you your experience on scrubbing in Safari. Here is a [link](https://www.cellstructureatlas.org/videos/1_10_Bbacteriovorus.mp4) to one of our videos using the native player. It seems like Safari has lost some of its scrubbing magic on my end. Is it the same for you?

**Comment by coiko** [comment]
Yep, definitely less magical for me, too.  (Lots of gaps/skips.)  Thanks @KianBadie!

**Comment by KianBadie** [comment]
@coiko Ah I see, that is unfortunate. The timing might work out because the new MacOS 15 version of safari does implement the chrome specific stuff we were using for scrubbing. I will keep you updated if we can get that working.

**Comment by coiko** [comment]
Awesome.  Thanks @KianBadie!

**Comment by KianBadie** [comment]
@coiko I spent some time looking into our current version of the scrubbing. While I was able to create a fix to the current flashing bug, I still couldn't help but worry about it happening again. The user being able to control the video and the way we capture frames while the video is playing makes it hard for me to be confident in how the behavior will turn out in all scenarios.

Because of that, it was hard for me to commit to that method without worry. So I spent some time on an alternate method that we have discussed before (I attached a gif below) where pre saved images are loaded from the host. I think it is a more traditional way of scrubbing video that has its benefits as well as its tradeoffs. The big benefits (so far, I am still in a prototype phase) are that it seems to be much easier to maintain, it is predictable since all the frames are saved for their predetermined times already, and scrubbing is determined by network vs what has played already. The main con is that the user has to fetch all the images, which might not be so bad if it is used on wifi/wired internet where a big network payload might not cost any finite data (like a mobile network). I got in touch with Tom about this idea and he said storing the images will not be a problem.

I can go over this in our meeting today, but I wanted to type it out here so its in our history. 

![demo](https://user-images.githubusercontent.com/18221058/149557178-e55d11f1-560c-4158-95ce-eef558443242.gif)

**Comment by KianBadie** [comment]
@coiko I tried things out with only using every other frame. The scrubbing still looks smooth, but I imagine a hands on demo would determine if it is satisfying enough. The biggest tradeoff is that since a bit can happen in 2 frames, the scrub is out of sync with the video. I have attached another gif below of what the scrubber shows for a given time, and what the video has for that same time (I did this by repeatedly clicking and releasing the seek bar). I still haven't experimented yet with webp images or seeing if there is a way to compress the images without loosing too much quality. I will keep you updated!

![demo](https://user-images.githubusercontent.com/18221058/149595339-d92abea2-d59d-4271-b27d-920d45e37a21.gif)
![fidget](https://user-images.githubusercontent.com/18221058/149595350-7b5a0baf-3807-459f-a788-16f2818f76c6.gif)

**Comment by KianBadie** [comment]
@coiko Which adobe program did you use again to make the videos? I know you used Adobe Media Encoder, but I forgot if you used a different program before that step.

**Comment by coiko** [comment]
Thanks so much @KianBadie!  Do you think the off-by-a-frame effect is too jarring as you're playing with it?

And good memory!  I used Adobe After Effects to make the movies, then Media Encoder to output.

**Comment by KianBadie** [comment]
@coiko I personally don't think it is jarring. I think I am still just a bit worried on how flashing is defined and how it relates to this. [For the highest level of conformance, web pages shouldn't have anything that flashes more than 3 times in a second in any given period](https://www.w3.org/TR/UNDERSTANDING-WCAG20/seizure-three-times.html). Seeking with every other frames, worst case scenario that I can think of, has the potential to flash a couple times if seeking lands on a scene change. This is because someone can seek from scene n to scene n+1, but because only every other frame is used scene n+2 is immediately shown because it is the actual frame at the seeked time. I'm not sure if this is over kill to consider, but it is something I would like to confirm first.

Ah I see! I ask because I tried to extract the frames from the video with no loss in quality but I still encountered some minor differences. The quality was on par but there were some difference in color (like minor differences in the shades of blue used). While minor, I did want to see which options your program had because I was wondering if extracting the frames from the source program would result in the best quality. I will be googling what the possibilities are for After Effects/Media Encoder!

Lastly, I will be emailing Tom a set of images to put on the testing site. Regardless of how many frames used and where the images come from, I want to see how things work out when the images are served from the actual server.

**Comment by coiko** [comment]
Thanks @KianBadie!  That all sounds great.  Thanks for checking up on these issues!

**Comment by KianBadie** [comment]
@coiko No problem!

**Comment by KianBadie** [comment]
@coiko My apologies for how long this has been taking. To be honest, I think I've been a little spooked with this feature ever since avoiding flashing has become a priority. While working on the feature this week I was still thinking of scenarios where flashing could occur, so I ended up creating an even more cautious approach than before. I attached a gif below of how it works.

Basically, I separated the video functionality from the scrub functionality. I detached scrubbing from the video timeline and when scrubbing ends the video player reverts back to where it was before scrubbing began. So it's less of a video scrubber and more of its own thing. That way, frames being out of sync from the video shouldn't be a problem since they are their own entities. What are your thoughts?

![demo](https://user-images.githubusercontent.com/18221058/150458571-1ceda2a7-d6b4-49b0-8b2d-5dc77ae88e5e.gif)
